File: 回測模組/strategy_modules/config_loader.py
"""
策略設定載入模組
負責載入和管理策略設定
"""

import os
import yaml
import logging
import pandas as pd
from datetime import datetime, time
from typing import Dict, Optional

logger = logging.getLogger(__name__)


class StrategyConfig:
    """策略設定管理器"""

    def __init__(self, config_path: str = "Bo_v2.yaml"):
        """
        初始化策略設定

        Args:
            config_path: 設定檔路徑
        """
        self.config_path = config_path
        self._config = self._load_config()

    def _load_config(self) -> Dict:
        """載入設定檔"""
        default_config = self._get_default_config()

        if not os.path.exists(self.config_path):
            logger.warning(f"設定檔不存在: {self.config_path}，使用預設值")
            return default_config

        try:
            with open(self.config_path, 'r', encoding='utf-8') as f:
                cfg = yaml.safe_load(f) or {}
        except Exception as exc:
            logger.warning(f"無法載入設定檔: {exc}，使用預設值")
            return default_config

        # 合併預設值和載入的設定
        strategy = cfg.get('strategy', {})

        # 更新設定值
        config = default_config.copy()

        # 時間設定
        config['entry_start_time'] = self._parse_time(
            strategy.get('entry_start_time', '09:09:00'),
            default_config['entry_start_time']
        )
        config['entry_cutoff_time'] = self._parse_time(
            strategy.get('entry_cutoff_time', '13:00:00'),
            default_config['entry_cutoff_time']
        )

        # 進場條件設定
        config['entry_cooldown'] = float(strategy.get('entry_cooldown', 30.0))
        config['ask_bid_ratio_threshold'] = float(strategy.get('ask_bid_ratio_threshold', 1.0))
        config['massive_matching_enabled'] = bool(strategy.get('massive_matching_enabled', True))
        config['massive_matching_amount'] = float(strategy.get('massive_matching_amount', 1000000.0))

        # 動態流動性門檻設定
        config['use_dynamic_liquidity_threshold'] = bool(strategy.get('use_dynamic_liquidity_threshold', False))
        config['dynamic_liquidity_threshold_file'] = str(strategy.get('dynamic_liquidity_threshold_file', 'daily_liquidity_threshold.parquet'))
        config['dynamic_liquidity_multiplier'] = float(strategy.get('dynamic_liquidity_multiplier', 0.004))  # 預設 0.4%
        config['dynamic_liquidity_threshold_cap'] = float(strategy.get('dynamic_liquidity_threshold_cap', 50_000_000))  # 上限 5000 萬
        config['dynamic_liquidity_threshold_df'] = None  # 初始化為 None，稍後載入

        config['ratio_entry_enabled'] = bool(strategy.get('ratio_entry_enabled', True))
        config['ratio_entry_threshold'] = float(strategy.get('ratio_entry_threshold', 3.0))
        config['ratio_column'] = str(strategy.get('ratio_column', 'ratio_15s_300s'))

        # *** 重要修改：漲幅限制從 7.4% 改為 8.5% ***
        config['price_change_limit_enabled'] = bool(strategy.get('price_change_limit_enabled', True))
        config['price_change_limit_pct'] = float(strategy.get('price_change_limit_pct', 8.5))  # 預設改為 8.5%

        # 區間漲幅濾網
        config['interval_pct_filter_enabled'] = bool(strategy.get('interval_pct_filter_enabled', False))
        minutes = int(strategy.get('interval_pct_minutes', 3))
        config['interval_pct_minutes'] = minutes if minutes in (2, 3, 5) else 3
        config['interval_pct_threshold'] = float(strategy.get('interval_pct_threshold', 4.0))

        # 進場 Buffer 機制
        config['entry_buffer_enabled'] = bool(strategy.get('entry_buffer_enabled', False))
        config['entry_buffer_milliseconds'] = int(strategy.get('entry_buffer_milliseconds', 20))

        # 平盤檢查設定
        config['above_open_check_enabled'] = bool(strategy.get('above_open_check_enabled', True))

        # 10分鐘低點檢查
        config['low_10min_filter_enabled'] = bool(strategy.get('low_10min_filter_enabled', False))
        config['low_10min_threshold'] = float(strategy.get('low_10min_threshold', 4.0))

        # 突破品質檢查設定（新策略）
        config['breakout_quality_check_enabled'] = bool(strategy.get('breakout_quality_check_enabled', False))  # 預設改為 False
        config['breakout_min_volume'] = int(strategy.get('breakout_min_volume', 10))
        config['breakout_min_ask_eat_ratio'] = float(strategy.get('breakout_min_ask_eat_ratio', 0.25))
        config['breakout_absolute_large_volume'] = int(strategy.get('breakout_absolute_large_volume', 50))

        # 小單推升過濾設定（舊策略）
        config['small_order_filter_enabled'] = bool(strategy.get('small_order_filter_enabled', False))  # 預設改為False
        config['small_order_check_trades'] = int(strategy.get('small_order_check_trades', 30))
        config['small_order_threshold'] = int(strategy.get('small_order_threshold', 3))
        config['tiny_order_threshold'] = int(strategy.get('tiny_order_threshold', 2))
        config['single_order_threshold'] = int(strategy.get('single_order_threshold', 1))
        config['small_order_ratio_limit'] = float(strategy.get('small_order_ratio_limit', 0.8))
        config['tiny_order_ratio_limit'] = float(strategy.get('tiny_order_ratio_limit', 0.5))
        config['single_order_ratio_limit'] = float(strategy.get('single_order_ratio_limit', 0.4))
        config['require_large_order_confirmation'] = bool(strategy.get('require_large_order_confirmation', False))
        config['large_order_threshold'] = int(strategy.get('large_order_threshold', 20))
        config['min_large_orders'] = int(strategy.get('min_large_orders', 2))

        # 重複進場設定
        config['reentry'] = bool(strategy.get('reentry', False))
        config['time_settings'] = str(strategy.get('time_settings', '1s'))

        # 輸出路徑設定
        config['output_path'] = str(strategy.get('output_path', 'D:/回測結果'))

        # Ratio 門檻遞增設定
        config['ratio_increase_after_loss_enabled'] = bool(strategy.get('ratio_increase_after_loss_enabled', False))
        config['ratio_increase_multiplier'] = float(strategy.get('ratio_increase_multiplier', 1.0))
        config['ratio_increase_min_threshold'] = float(strategy.get('ratio_increase_min_threshold', 6.0))

        # 停損/停利參數
        config['strategy_b_stop_loss_ticks_small'] = int(strategy.get('strategy_b_stop_loss_ticks_small', 4))
        config['strategy_b_stop_loss_ticks_large'] = int(strategy.get('strategy_b_stop_loss_ticks_large', 3))
        config['momentum_stop_loss_seconds'] = float(strategy.get('momentum_stop_loss_seconds', 15.0))
        config['momentum_stop_loss_seconds_extended'] = float(strategy.get('momentum_stop_loss_seconds_extended', 30.0))
        config['momentum_stop_loss_amount'] = float(strategy.get('momentum_stop_loss_amount', 5000000.0))
        config['momentum_extended_capital_threshold'] = float(strategy.get('momentum_extended_capital_threshold', 50000000000))
        config['momentum_extended_price_min'] = float(strategy.get('momentum_extended_price_min', 100.0))
        config['momentum_extended_price_max'] = float(strategy.get('momentum_extended_price_max', 500.0))

        # 移動停利設定
        config['trailing_stop'] = strategy.get('trailing_stop', {
            'enabled': False,
            'levels': [],
            'entry_price_protection': True
        })

        # 載入動態流動性門檻資料（如果啟用）
        if config['use_dynamic_liquidity_threshold']:
            try:
                threshold_file = config['dynamic_liquidity_threshold_file']
                # 如果是相對路徑，則相對於config檔案所在目錄
                if not os.path.isabs(threshold_file):
                    config_dir = os.path.dirname(os.path.abspath(self.config_path))
                    threshold_file = os.path.join(config_dir, threshold_file)

                if os.path.exists(threshold_file):
                    config['dynamic_liquidity_threshold_df'] = pd.read_parquet(threshold_file)
                    logger.info(f"✓ 已載入動態流動性門檻檔案: {threshold_file}")
                    logger.info(f"  資料範圍: {config['dynamic_liquidity_threshold_df'].index.min()} ~ {config['dynamic_liquidity_threshold_df'].index.max()}")
                    logger.info(f"  包含股票數: {len(config['dynamic_liquidity_threshold_df'].columns)}")
                else:
                    logger.warning(f"⚠ 動態流動性門檻檔案不存在: {threshold_file}")
                    logger.warning(f"  將使用固定門檻 massive_matching_amount")
                    config['use_dynamic_liquidity_threshold'] = False
            except Exception as e:
                logger.error(f"❌ 載入動態流動性門檻檔案失敗: {e}")
                logger.warning(f"  將使用固定門檻 massive_matching_amount")
                config['use_dynamic_liquidity_threshold'] = False
                config['dynamic_liquidity_threshold_df'] = None

        return config

    def _get_default_config(self) -> Dict:
        """取得預設設定"""
        return {
            # 時間設定
            'entry_start_time': time(9, 9, 0),
            'entry_cutoff_time': time(13, 0, 0),
            'entry_cooldown': 30.0,

            # 進場條件
            'ask_bid_ratio_threshold': 1.0,
            'massive_matching_enabled': True,
            'massive_matching_amount': 1000000.0,
            'use_dynamic_liquidity_threshold': False,
            'dynamic_liquidity_threshold_file': 'daily_liquidity_threshold.parquet',
            'dynamic_liquidity_multiplier': 0.004,  # 預設 0.4%
            'dynamic_liquidity_threshold_cap': 50_000_000,  # 上限 5000 萬
            'dynamic_liquidity_threshold_df': None,
            'ratio_entry_enabled': True,
            'ratio_entry_threshold': 3.0,
            'ratio_column': 'ratio_15s_300s',

            # *** 漲幅限制改為 8.5% ***
            'price_change_limit_enabled': True,
            'price_change_limit_pct': 8.5,  # 從 7.4% 改為 8.5%

            # 區間漲幅濾網
            'interval_pct_filter_enabled': False,
            'interval_pct_minutes': 3,
            'interval_pct_threshold': 4.0,

            # 進場 Buffer
            'entry_buffer_enabled': False,
            'entry_buffer_milliseconds': 20,

            # 10分鐘低點檢查
            'low_10min_filter_enabled': False,
            'low_10min_threshold': 4.0,

            # 重複進場
            'reentry': False,
            'time_settings': '1s',

            # 輸出路徑
            'output_path': 'D:/回測結果',

            # Ratio 門檻遞增
            'ratio_increase_after_loss_enabled': False,
            'ratio_increase_multiplier': 1.0,
            'ratio_increase_min_threshold': 6.0,

            # 停損/停利參數
            'strategy_b_stop_loss_ticks_small': 4,
            'strategy_b_stop_loss_ticks_large': 3,
            'momentum_stop_loss_seconds': 15.0,
            'momentum_stop_loss_seconds_extended': 30.0,
            'momentum_stop_loss_amount': 5000000.0,
            'momentum_extended_capital_threshold': 50000000000,
            'momentum_extended_price_min': 100.0,
            'momentum_extended_price_max': 500.0
        }

    def _parse_time(self, value: str, fallback: time) -> time:
        """解析時間字串"""
        try:
            return datetime.strptime(str(value), '%H:%M:%S').time()
        except Exception:
            return fallback

    def get_entry_config(self) -> Dict:
        """取得進場相關設定"""
        return {
            'entry_start_time': self._config['entry_start_time'],
            'entry_cutoff_time': self._config['entry_cutoff_time'],
            'entry_cooldown': self._config['entry_cooldown'],
            'above_open_check_enabled': self._config.get('above_open_check_enabled', True),  # 新增平盤檢查設定
            'ask_bid_ratio_threshold': self._config['ask_bid_ratio_threshold'],
            'massive_matching_enabled': self._config['massive_matching_enabled'],
            'massive_matching_amount': self._config['massive_matching_amount'],
            'use_dynamic_liquidity_threshold': self._config['use_dynamic_liquidity_threshold'],
            'dynamic_liquidity_threshold_file': self._config['dynamic_liquidity_threshold_file'],
            'dynamic_liquidity_multiplier': self._config['dynamic_liquidity_multiplier'],
            'dynamic_liquidity_threshold_cap': self._config['dynamic_liquidity_threshold_cap'],
            'dynamic_liquidity_threshold_df': self._config['dynamic_liquidity_threshold_df'],
            'ratio_entry_enabled': self._config['ratio_entry_enabled'],
            'ratio_entry_threshold': self._config['ratio_entry_threshold'],
            'ratio_column': self._config['ratio_column'],
            'price_change_limit_enabled': self._config['price_change_limit_enabled'],
            'price_change_limit_pct': self._config['price_change_limit_pct'],
            'interval_pct_filter_enabled': self._config['interval_pct_filter_enabled'],
            'interval_pct_minutes': self._config['interval_pct_minutes'],
            'interval_pct_threshold': self._config['interval_pct_threshold'],
            'entry_buffer_enabled': self._config['entry_buffer_enabled'],
            'entry_buffer_milliseconds': self._config['entry_buffer_milliseconds'],
            'low_10min_filter_enabled': self._config.get('low_10min_filter_enabled', False),
            'low_10min_threshold': self._config.get('low_10min_threshold', 4.0),
            # 突破品質檢查設定
            'breakout_quality_check_enabled': self._config.get('breakout_quality_check_enabled', False),
            'breakout_min_volume': self._config.get('breakout_min_volume', 10),
            'breakout_min_ask_eat_ratio': self._config.get('breakout_min_ask_eat_ratio', 0.25),
            'breakout_absolute_large_volume': self._config.get('breakout_absolute_large_volume', 50),
            # 小單過濾設定
            'small_order_filter_enabled': self._config.get('small_order_filter_enabled', False),
            'small_order_check_trades': self._config.get('small_order_check_trades', 30),
            'small_order_threshold': self._config.get('small_order_threshold', 3),
            'tiny_order_threshold': self._config.get('tiny_order_threshold', 2),
            'single_order_threshold': self._config.get('single_order_threshold', 1),
            'small_order_ratio_limit': self._config.get('small_order_ratio_limit', 0.8),
            'tiny_order_ratio_limit': self._config.get('tiny_order_ratio_limit', 0.5),
            'single_order_ratio_limit': self._config.get('single_order_ratio_limit', 0.4),
            'require_large_order_confirmation': self._config.get('require_large_order_confirmation', False),
            'large_order_threshold': self._config.get('large_order_threshold', 20),
            'min_large_orders': self._config.get('min_large_orders', 2)
        }

    def get_exit_config(self) -> Dict:
        """取得出場相關設定"""
        return {
            'strategy_b_stop_loss_ticks_small': self._config['strategy_b_stop_loss_ticks_small'],
            'strategy_b_stop_loss_ticks_large': self._config['strategy_b_stop_loss_ticks_large'],
            'momentum_stop_loss_seconds': self._config['momentum_stop_loss_seconds'],
            'momentum_stop_loss_seconds_extended': self._config['momentum_stop_loss_seconds_extended'],
            'momentum_stop_loss_amount': self._config['momentum_stop_loss_amount'],
            'momentum_extended_capital_threshold': self._config['momentum_extended_capital_threshold'],
            'momentum_extended_price_min': self._config['momentum_extended_price_min'],
            'momentum_extended_price_max': self._config['momentum_extended_price_max'],
            # 移動停利配置
            'trailing_stop': self._config.get('trailing_stop', {})
        }

    def get_reentry_config(self) -> Dict:
        """取得重複進場相關設定"""
        return {
            'reentry': self._config['reentry'],
            'time_settings': self._config['time_settings'],
            'ratio_increase_after_loss_enabled': self._config['ratio_increase_after_loss_enabled'],
            'ratio_increase_multiplier': self._config['ratio_increase_multiplier'],
            'ratio_increase_min_threshold': self._config['ratio_increase_min_threshold']
        }

    def get_all_config(self) -> Dict:
        """取得所有設定"""
        return self._config.copy()

File: 回測模組/CSharp/BacktestModule/Strategy/ConfigLoader.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Globalization;
using YamlDotNet.Serialization;
using YamlDotNet.Serialization.NamingConventions;

namespace BacktestModule.Strategy
{
    /// <summary>
    /// YAML-based strategy configuration loader.
    /// Mirrors Python StrategyConfig: loads YAML, merges with defaults, provides typed access.
    /// </summary>
    public class StrategyConfig
    {
        private readonly Dictionary<string, object> _config;
        private readonly string _configPath;

        public StrategyConfig(string configPath = "Bo_v2.yaml")
        {
            _configPath = configPath;
            _config = LoadConfig();
        }

        private Dictionary<string, object> LoadConfig()
        {
            var config = GetDefaultConfig();

            if (!File.Exists(_configPath))
            {
                Console.WriteLine($"[WARNING] Config file not found: {_configPath}, using defaults.");
                return config;
            }

            try
            {
                var yaml = File.ReadAllText(_configPath);
                var deserializer = new DeserializerBuilder()
                    .WithNamingConvention(UnderscoredNamingConvention.Instance)
                    .Build();
                var root = deserializer.Deserialize<Dictionary<string, object>>(yaml);

                if (root == null) return config;

                // Extract the 'strategy' section
                Dictionary<string, object> strategy = new Dictionary<string, object>();
                if (root.ContainsKey("strategy") && root["strategy"] is Dictionary<object, object> rawStrategy)
                {
                    foreach (var kv in rawStrategy)
                        strategy[kv.Key?.ToString() ?? ""] = kv.Value;
                }

                // Time settings
                config["entry_start_time"] = ParseTime(
                    GetString(strategy, "entry_start_time", "09:05:00"),
                    (TimeSpan)config["entry_start_time"]);
                config["entry_cutoff_time"] = ParseTime(
                    GetString(strategy, "entry_cutoff_time", "13:00:00"),
                    (TimeSpan)config["entry_cutoff_time"]);

                // Entry conditions
                config["entry_cooldown"] = GetDouble(strategy, "entry_cooldown", 30.0);
                config["ask_bid_ratio_threshold"] = GetDouble(strategy, "ask_bid_ratio_threshold", 1.0);
                config["massive_matching_enabled"] = GetBool(strategy, "massive_matching_enabled", true);
                config["massive_matching_amount"] = GetDouble(strategy, "massive_matching_amount", 1000000.0);

                // Dynamic liquidity threshold
                config["use_dynamic_liquidity_threshold"] = GetBool(strategy, "use_dynamic_liquidity_threshold", false);
                config["dynamic_liquidity_threshold_file"] = GetString(strategy, "dynamic_liquidity_threshold_file", "daily_liquidity_threshold.parquet");
                config["dynamic_liquidity_multiplier"] = GetDouble(strategy, "dynamic_liquidity_multiplier", 0.004);
                config["dynamic_liquidity_threshold_cap"] = GetDouble(strategy, "dynamic_liquidity_threshold_cap", 50_000_000);

                config["ratio_entry_enabled"] = GetBool(strategy, "ratio_entry_enabled", true);
                config["ratio_entry_threshold"] = GetDouble(strategy, "ratio_entry_threshold", 3.0);
                config["ratio_column"] = GetString(strategy, "ratio_column", "ratio_15s_300s");

                // Price change limit
                config["price_change_limit_enabled"] = GetBool(strategy, "price_change_limit_enabled", true);
                config["price_change_limit_pct"] = GetDouble(strategy, "price_change_limit_pct", 8.5);

                // Interval percentage filter
                config["interval_pct_filter_enabled"] = GetBool(strategy, "interval_pct_filter_enabled", false);
                int minutes = GetInt(strategy, "interval_pct_minutes", 3);
                config["interval_pct_minutes"] = (minutes == 2 || minutes == 3 || minutes == 5) ? minutes : 3;
                config["interval_pct_threshold"] = GetDouble(strategy, "interval_pct_threshold", 4.0);

                // Entry buffer
                config["entry_buffer_enabled"] = GetBool(strategy, "entry_buffer_enabled", false);
                config["entry_buffer_milliseconds"] = GetInt(strategy, "entry_buffer_milliseconds", 20);

                // Above open check
                config["above_open_check_enabled"] = GetBool(strategy, "above_open_check_enabled", true);

                // 10-minute low filter
                config["low_10min_filter_enabled"] = GetBool(strategy, "low_10min_filter_enabled", false);
                config["low_10min_threshold"] = GetDouble(strategy, "low_10min_threshold", 4.0);

                // Breakout quality
                config["breakout_quality_check_enabled"] = GetBool(strategy, "breakout_quality_check_enabled", false);
                config["breakout_min_volume"] = GetInt(strategy, "breakout_min_volume", 10);
                config["breakout_min_ask_eat_ratio"] = GetDouble(strategy, "breakout_min_ask_eat_ratio", 0.25);
                config["breakout_absolute_large_volume"] = GetInt(strategy, "breakout_absolute_large_volume", 50);

                // Small order filter
                config["small_order_filter_enabled"] = GetBool(strategy, "small_order_filter_enabled", false);
                config["small_order_check_trades"] = GetInt(strategy, "small_order_check_trades", 30);
                config["small_order_threshold"] = GetInt(strategy, "small_order_threshold", 3);
                config["tiny_order_threshold"] = GetInt(strategy, "tiny_order_threshold", 2);
                config["single_order_threshold"] = GetInt(strategy, "single_order_threshold", 1);
                config["small_order_ratio_limit"] = GetDouble(strategy, "small_order_ratio_limit", 0.8);
                config["tiny_order_ratio_limit"] = GetDouble(strategy, "tiny_order_ratio_limit", 0.5);
                config["single_order_ratio_limit"] = GetDouble(strategy, "single_order_ratio_limit", 0.4);
                config["require_large_order_confirmation"] = GetBool(strategy, "require_large_order_confirmation", false);
                config["large_order_threshold"] = GetInt(strategy, "large_order_threshold", 20);
                config["min_large_orders"] = GetInt(strategy, "min_large_orders", 2);

                // Reentry
                config["reentry"] = GetBool(strategy, "reentry", false);
                config["time_settings"] = GetString(strategy, "time_settings", "1s");

                // Output path
                config["output_path"] = GetString(strategy, "output_path", @"D:/回測結果");

                // Ratio increase after loss
                config["ratio_increase_after_loss_enabled"] = GetBool(strategy, "ratio_increase_after_loss_enabled", false);
                config["ratio_increase_multiplier"] = GetDouble(strategy, "ratio_increase_multiplier", 1.0);
                config["ratio_increase_min_threshold"] = GetDouble(strategy, "ratio_increase_min_threshold", 6.0);

                // Stop-loss parameters
                config["strategy_b_stop_loss_ticks_small"] = GetInt(strategy, "strategy_b_stop_loss_ticks_small", 4);
                config["strategy_b_stop_loss_ticks_large"] = GetInt(strategy, "strategy_b_stop_loss_ticks_large", 3);
                config["momentum_stop_loss_seconds"] = GetDouble(strategy, "momentum_stop_loss_seconds", 15.0);
                config["momentum_stop_loss_seconds_extended"] = GetDouble(strategy, "momentum_stop_loss_seconds_extended", 30.0);
                config["momentum_stop_loss_amount"] = GetDouble(strategy, "momentum_stop_loss_amount", 5000000.0);
                config["momentum_extended_capital_threshold"] = GetDouble(strategy, "momentum_extended_capital_threshold", 50000000000);
                config["momentum_extended_price_min"] = GetDouble(strategy, "momentum_extended_price_min", 100.0);
                config["momentum_extended_price_max"] = GetDouble(strategy, "momentum_extended_price_max", 500.0);

                // Trailing stop configuration
                config["trailing_stop"] = ParseTrailingStopConfig(strategy);
            }
            catch (Exception ex)
            {
                Console.WriteLine($"[WARNING] Failed to load config: {ex.Message}, using defaults.");
            }

            return config;
        }

        private Dictionary<string, object> ParseTrailingStopConfig(Dictionary<string, object> strategy)
        {
            var result = new Dictionary<string, object>
            {
                ["enabled"] = false,
                ["levels"] = new List<Dictionary<string, object>>(),
                ["entry_price_protection"] = true
            };

            if (!strategy.ContainsKey("trailing_stop")) return result;

            if (strategy["trailing_stop"] is Dictionary<object, object> tsRaw)
            {
                var ts = new Dictionary<string, object>();
                foreach (var kv in tsRaw)
                    ts[kv.Key?.ToString() ?? ""] = kv.Value;

                result["enabled"] = GetBool(ts, "enabled", false);
                result["entry_price_protection"] = GetBool(ts, "entry_price_protection", true);

                var levels = new List<Dictionary<string, object>>();
                if (ts.ContainsKey("levels") && ts["levels"] is List<object> rawLevels)
                {
                    foreach (var rawLevel in rawLevels)
                    {
                        if (rawLevel is Dictionary<object, object> levelDict)
                        {
                            var level = new Dictionary<string, object>();
                            foreach (var kv in levelDict)
                                level[kv.Key?.ToString() ?? ""] = kv.Value;

                            levels.Add(new Dictionary<string, object>
                            {
                                ["name"] = GetString(level, "name", ""),
                                ["field"] = GetString(level, "field", ""),
                                ["exit_ratio"] = GetDouble(level, "exit_ratio", 0.333)
                            });
                        }
                    }
                }
                result["levels"] = levels;
            }

            return result;
        }

        private Dictionary<string, object> GetDefaultConfig()
        {
            return new Dictionary<string, object>
            {
                ["entry_start_time"] = new TimeSpan(9, 9, 0),
                ["entry_cutoff_time"] = new TimeSpan(13, 0, 0),
                ["entry_cooldown"] = 30.0,
                ["ask_bid_ratio_threshold"] = 1.0,
                ["massive_matching_enabled"] = true,
                ["massive_matching_amount"] = 1000000.0,
                ["use_dynamic_liquidity_threshold"] = false,
                ["dynamic_liquidity_threshold_file"] = "daily_liquidity_threshold.parquet",
                ["dynamic_liquidity_multiplier"] = 0.004,
                ["dynamic_liquidity_threshold_cap"] = 50_000_000.0,
                ["ratio_entry_enabled"] = true,
                ["ratio_entry_threshold"] = 3.0,
                ["ratio_column"] = "ratio_15s_300s",
                ["price_change_limit_enabled"] = true,
                ["price_change_limit_pct"] = 8.5,
                ["interval_pct_filter_enabled"] = false,
                ["interval_pct_minutes"] = 3,
                ["interval_pct_threshold"] = 4.0,
                ["entry_buffer_enabled"] = false,
                ["entry_buffer_milliseconds"] = 20,
                ["above_open_check_enabled"] = true,
                ["low_10min_filter_enabled"] = false,
                ["low_10min_threshold"] = 4.0,
                ["breakout_quality_check_enabled"] = false,
                ["breakout_min_volume"] = 10,
                ["breakout_min_ask_eat_ratio"] = 0.25,
                ["breakout_absolute_large_volume"] = 50,
                ["small_order_filter_enabled"] = false,
                ["small_order_check_trades"] = 30,
                ["small_order_threshold"] = 3,
                ["tiny_order_threshold"] = 2,
                ["single_order_threshold"] = 1,
                ["small_order_ratio_limit"] = 0.8,
                ["tiny_order_ratio_limit"] = 0.5,
                ["single_order_ratio_limit"] = 0.4,
                ["require_large_order_confirmation"] = false,
                ["large_order_threshold"] = 20,
                ["min_large_orders"] = 2,
                ["reentry"] = false,
                ["time_settings"] = "1s",
                ["output_path"] = @"D:/回測結果",
                ["ratio_increase_after_loss_enabled"] = false,
                ["ratio_increase_multiplier"] = 1.0,
                ["ratio_increase_min_threshold"] = 6.0,
                ["strategy_b_stop_loss_ticks_small"] = 4,
                ["strategy_b_stop_loss_ticks_large"] = 3,
                ["momentum_stop_loss_seconds"] = 15.0,
                ["momentum_stop_loss_seconds_extended"] = 30.0,
                ["momentum_stop_loss_amount"] = 5000000.0,
                ["momentum_extended_capital_threshold"] = 50000000000.0,
                ["momentum_extended_price_min"] = 100.0,
                ["momentum_extended_price_max"] = 500.0,
                ["trailing_stop"] = new Dictionary<string, object>
                {
                    ["enabled"] = false,
                    ["levels"] = new List<Dictionary<string, object>>(),
                    ["entry_price_protection"] = true
                }
            };
        }

        private static TimeSpan ParseTime(string value, TimeSpan fallback)
        {
            if (TimeSpan.TryParseExact(value, @"hh\:mm\:ss", CultureInfo.InvariantCulture, out var ts))
                return ts;
            if (TimeSpan.TryParse(value, out ts))
                return ts;
            return fallback;
        }

        // ===== Helper methods for extracting typed values from YAML dict =====

        private static string GetString(Dictionary<string, object> d, string key, string defaultVal)
        {
            if (d.TryGetValue(key, out var v) && v != null)
                return v.ToString() ?? defaultVal;
            return defaultVal;
        }

        private static double GetDouble(Dictionary<string, object> d, string key, double defaultVal)
        {
            if (d.TryGetValue(key, out var v) && v != null)
            {
                if (v is double dv) return dv;
                if (v is float fv) return fv;
                if (v is int iv) return iv;
                if (v is long lv) return lv;
                if (double.TryParse(v.ToString(), NumberStyles.Any, CultureInfo.InvariantCulture, out var parsed))
                    return parsed;
            }
            return defaultVal;
        }

        private static int GetInt(Dictionary<string, object> d, string key, int defaultVal)
        {
            if (d.TryGetValue(key, out var v) && v != null)
            {
                if (v is int iv) return iv;
                if (v is long lv) return (int)lv;
                if (v is double dv) return (int)dv;
                if (int.TryParse(v.ToString(), out var parsed))
                    return parsed;
            }
            return defaultVal;
        }

        private static bool GetBool(Dictionary<string, object> d, string key, bool defaultVal)
        {
            if (d.TryGetValue(key, out var v) && v != null)
            {
                if (v is bool bv) return bv;
                if (bool.TryParse(v.ToString(), out var parsed))
                    return parsed;
                // YAML "true"/"false" strings
                string s = v.ToString()?.Trim().ToLowerInvariant() ?? "";
                if (s == "true" || s == "yes" || s == "1") return true;
                if (s == "false" || s == "no" || s == "0") return false;
            }
            return defaultVal;
        }

        // ===== Public API: get config sections =====

        public Dictionary<string, object> GetEntryConfig()
        {
            var keys = new[]
            {
                "entry_start_time", "entry_cutoff_time", "entry_cooldown",
                "above_open_check_enabled", "ask_bid_ratio_threshold",
                "massive_matching_enabled", "massive_matching_amount",
                "use_dynamic_liquidity_threshold", "dynamic_liquidity_threshold_file",
                "dynamic_liquidity_multiplier", "dynamic_liquidity_threshold_cap",
                "ratio_entry_enabled", "ratio_entry_threshold", "ratio_column",
                "price_change_limit_enabled", "price_change_limit_pct",
                "interval_pct_filter_enabled", "interval_pct_minutes", "interval_pct_threshold",
                "entry_buffer_enabled", "entry_buffer_milliseconds",
                "low_10min_filter_enabled", "low_10min_threshold",
                "breakout_quality_check_enabled", "breakout_min_volume",
                "breakout_min_ask_eat_ratio", "breakout_absolute_large_volume",
                "small_order_filter_enabled", "small_order_check_trades",
                "small_order_threshold", "tiny_order_threshold", "single_order_threshold",
                "small_order_ratio_limit", "tiny_order_ratio_limit", "single_order_ratio_limit",
                "require_large_order_confirmation", "large_order_threshold", "min_large_orders",
                "ratio_increase_after_loss_enabled", "ratio_increase_multiplier",
                "ratio_increase_min_threshold"
            };
            return ExtractKeys(keys);
        }

        public Dictionary<string, object> GetExitConfig()
        {
            var keys = new[]
            {
                "strategy_b_stop_loss_ticks_small", "strategy_b_stop_loss_ticks_large",
                "momentum_stop_loss_seconds", "momentum_stop_loss_seconds_extended",
                "momentum_stop_loss_amount", "momentum_extended_capital_threshold",
                "momentum_extended_price_min", "momentum_extended_price_max",
                "trailing_stop"
            };
            return ExtractKeys(keys);
        }

        public Dictionary<string, object> GetReentryConfig()
        {
            var keys = new[]
            {
                "reentry", "time_settings",
                "ratio_increase_after_loss_enabled",
                "ratio_increase_multiplier", "ratio_increase_min_threshold"
            };
            return ExtractKeys(keys);
        }

        public Dictionary<string, object> GetAllConfig()
        {
            return new Dictionary<string, object>(_config);
        }

        /// <summary>
        /// Direct access to underlying config for parameter overrides.
        /// </summary>
        public Dictionary<string, object> RawConfig => _config;

        private Dictionary<string, object> ExtractKeys(string[] keys)
        {
            var result = new Dictionary<string, object>();
            foreach (var key in keys)
            {
                if (_config.ContainsKey(key))
                    result[key] = _config[key];
            }
            return result;
        }
    }
}
